File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

#define ALL(X) X.begin(), X.end()
using namespace std;

vector<int> res;
vector<pair<int,int>> adj[50001];

int intensity[50001];
bool isSummit[50001];

void go(vector<int>& gates) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    vector<pair<int, int>> temp;
    memset(intensity, -1, sizeof(intensity));
    
    for(auto g : gates) {
        pq.push({0, g});
        intensity[g] = 0;
    }
    
    while(!pq.empty()) {
        int cost = pq.top().first;
        int nowV = pq.top().second;
        pq.pop();
    
        if(cost > intensity[nowV]) continue;
        
        if(isSummit[nowV]) {
            temp.push_back({cost, nowV});
            continue;
        }
    
        for(auto p : adj[nowV]) {
            int weight = p.first;
            int to = p.second;
            
            if(intensity[to] == -1 || max(cost, weight) < intensity[to]) {
                intensity[to] = max(cost, weight);
                pq.push({intensity[to], to});
            }
        }
    }
    
    sort(ALL(temp));
    res.push_back(temp[0].second);
    res.push_back(temp[0].first);
}

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {
    for (auto e : paths) {
        adj[e[0]].push_back({e[2], e[1]});
        adj[e[1]].push_back({e[2], e[0]});
    }
    
    for (auto s : summits) {
        isSummit[s] = true;
    }
    go(gates);
    return res;
}

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.java
import java.util.*;
import java.io.*;

class Solution {
    class Node {
        int to, intensity;
        Node(int to, int intensity) {
            this.to = to;
            this.intensity = intensity;
        }
    }
    
    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        List<Node>[] graph = new ArrayList[n+1];
        
        for(int i=1; i<=n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for(int[] path : paths) {
            int from = path[0];
            int to = path[1];
            int intensity = path[2];
            
            graph[from].add(new Node(to, intensity));
            graph[to].add(new Node(from, intensity));
        }
        
        Queue<Node> pq = new PriorityQueue<>((a, b) -> a.intensity - b.intensity);
        int[] distance = new int[n+1];
        Arrays.fill(distance, Integer.MAX_VALUE);
       
        for(int gate : gates) {
            distance[gate] = 0;
            pq.add(new Node(gate, 0));
        }
        
        boolean[] issummit = new boolean[n+1];
        for(int summit : summits) {
            issummit[summit] = true;
        }
        
        while(!pq.isEmpty()) {
            Node cur = pq.poll();
            int curnode = cur.to;
            int curintensity = cur.intensity;
            
            if(curintensity > distance[curnode]) continue;
            if(issummit[curnode]) continue;
            
            for(Node nxt : graph[curnode]) {
                int maxintensity = Math.max(curintensity, nxt.intensity);
                if(maxintensity < distance[nxt.to]) {
                    distance[nxt.to] = maxintensity;
                    pq.add(new Node(nxt.to, maxintensity));
                }
            }
        }
        
        Arrays.sort(summits);
        
        int summitnumber = -1;
        int minintensity = Integer.MAX_VALUE;
        
        for(int summit : summits) {
            if(distance[summit] < minintensity) {
                summitnumber = summit;
                minintensity = distance[summit];
            }
        }
        
        return new int[]{summitnumber, minintensity};
    }
}